### PR TITLE
input: pause input when storage chunks reach their limit.

### PR DESCRIPTION
**Pull request by pwhelan**
Add an option to pause input when storage chunks reach their limit.

----

**Testing**
- [x] Example configuration file for the change
- [x] Debug log output from testing the change
- [x] Attached [Valgrind](https://valgrind.org/docs/manual/quick-start.html) output that shows no leaks or memory corruption was found

**Documentation**
- [ ] Documentation required for this feature

configuration example: 
```
[SERVICE]
    storage.type filesystem
    storage.max_chunks_pause on
    storage.path /tmp/chunks

[INPUT]
    Name                  dummy
    storage.type          filesystem

[OUTPUT]
    Name stdout
    Match *
```

----

Fluent Bit is licensed under Apache 2.0, by submitting this pull request I understand that this code will be released under the terms of that license.

### CODE REVIEW

**Comment by pwhelan** [comment]
Here is a debug log of a run with this configuration:

```ini
[SERVICE]
    storage.type filesystem
    storage.max_chunks_pause on
    storage.path /tmp/chunks

[INPUT]
    Name                  dummy
    Rate                  260
    storage.type          filesystem

[OUTPUT]
    Name  stdout
    Match *

[OUTPUT]
    Name  exit
    Match *

```

[debug.log](https://github.com/fluent/fluent-bit/files/7010675/debug.log)

**Comment by pwhelan** [comment]
Valgrind Log generated using the following configuration:

```
[SERVICE]
    flush 1
    storage.type filesystem
    storage.max_chunks_up 1
    storage.max_chunks_pause on
    storage.path /tmp/chunks

[INPUT]
    Name                  dummy
    Rate                  1
    storage.type          filesystem

[OUTPUT]
    Name  stdout
    Match *

[OUTPUT]
    Name  tcp
    Host  does.not.exist
    Match *

[OUTPUT]
    Name  exit
    Match *
```

[valgrind.log](https://github.com/fluent/fluent-bit/files/7010714/valgrind.log)

File: src/flb_input_chunk.c [hunk 1]
```diff
@@ -788,16 +804,30 @@ size_t flb_input_chunk_set_limits(struct flb_input_instance *in)
      * After the adjustments, validate if the plugin is overlimit or paused
      * and perform further adjustments.
      */
-    if (flb_input_chunk_is_overlimit(in) == FLB_FALSE &&
-        flb_input_buf_paused(in) && in->config->is_running == FLB_TRUE &&
```
Comment (edsiper) [inline src/flb_input_chunk.c:792]: why flb_input_buf_paused() was removed ?

File: src/flb_input_chunk.c [hunk 2]
```diff
@@ -788,16 +804,30 @@ size_t flb_input_chunk_set_limits(struct flb_input_instance *in)
...
-        in->config->is_ingestion_active == FLB_TRUE) {
+    if (flb_input_chunk_is_mem_overlimit(in) == FLB_FALSE &&
+        in->config->is_running == FLB_TRUE &&
+        in->config->is_ingestion_active == FLB_TRUE &&
+        in->mem_buf_status == FLB_INPUT_PAUSED) {
         in->mem_buf_status = FLB_INPUT_RUNNING;
         if (in->p->cb_resume) {
             in->p->cb_resume(in->context, in->config);
-            flb_info("[input] %s resume (mem buf overlimit)",
+            flb_warn("[input] %s resume (mem buf overlimit)",
```
Comment (edsiper) [inline src/flb_input_chunk.c:814]: warning "warns" the user, we just need to inform him everything is ok

**Review by edsiper — CHANGES_REQUESTED** [review]

File: src/flb_input_chunk.c (hunk 1 above)
Comment (pwhelan) [reply src/flb_input_chunk.c:792]: I changed it for a specific reference to a check to in->mem_buf_status, that way pausing for mem_buf and storage are checked separately. To add back in the flb_input_buf_paused() call I could refactor the entire code to check the pause first then check which paused it either memory or storage.

**Comment by pwhelan** [comment]
Superseded by #4077 .

**Pull request closed without merging**